File: spec/requests/api/v1/auth_spec.rb
require "rails_helper"

RSpec.describe "Auth", type: :request do
  describe "POST /api/v1/auth/register" do
    it "creates a user and returns a jwt plus uuid user id" do
      post "/api/v1/auth/register",
           params: { username: "Alice", password: "secret123" },
           as: :json

      expect(response).to have_http_status(:created)
      user_id = json.dig("data", "user", "id")
      expect(user_id).to match(
        /\A[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\z/i
      )
      expect(user_id.length).to eq(36)
      expect(json.dig("data", "user", "username")).to eq("alice")
      expect(json.dig("data", "token")).to be_present
      expect(json.dig("data", "user")).not_to have_key("email")
      expect(User.find(user_id).username).to eq("alice")
    end

    it "ignores email and does not persist it" do
      post "/api/v1/auth/register",
           params: { username: "bob", password: "secret123", email: "bob@example.com" },
           as: :json

      expect(response).to have_http_status(:created)
      expect(User.column_names).not_to include("email")
      expect(json.dig("data", "user")).not_to have_key("email")
    end

    it "rejects a duplicate username regardless of case" do
      create(:user, username: "alice", password: "secret123")

      post "/api/v1/auth/register",
           params: { username: "ALICE", password: "secret123" },
           as: :json

      expect(response).to have_http_status(:unprocessable_content)
      expect(json.dig("error", "code")).to eq("validation_error")
    end

    it "rejects a password shorter than 8 characters" do
      post "/api/v1/auth/register",
           params: { username: "alice", password: "short" },
           as: :json

      expect(response).to have_http_status(:unprocessable_content)
      expect(json.dig("error", "code")).to eq("validation_error")
    end
  end

  describe "POST /api/v1/auth/login" do
    before { create(:user, username: "alice", password: "secret123") }

    it "returns a jwt for the correct password" do
      post "/api/v1/auth/login",
           params: { username: "ALICE", password: "secret123" },
           as: :json

      expect(response).to have_http_status(:ok)
      expect(json.dig("data", "token")).to be_present
      expect(json.dig("data", "user", "username")).to eq("alice")
      expect(json.dig("data", "user", "id").length).to eq(36)
    end

    it "returns invalid_credentials for a wrong password" do
      post "/api/v1/auth/login",
           params: { username: "alice", password: "wrong-password" },
           as: :json

      expect(response).to have_http_status(:unauthorized)
      expect(json.dig("error", "code")).to eq("invalid_credentials")
      expect(json.dig("error", "request_id")).to be_present
    end

    it "returns invalid_credentials for an unknown username" do
      post "/api/v1/auth/login",
           params: { username: "nobody", password: "secret123" },
           as: :json

      expect(response).to have_http_status(:unauthorized)
      expect(json.dig("error", "code")).to eq("invalid_credentials")
    end
  end

  describe "absent session endpoints" do
    it "does not expose DELETE /api/v1/auth/logout" do
      delete "/api/v1/auth/logout"
      expect(response).to have_http_status(:not_found)
    end

    it "does not expose /api/v1/sessions" do
      get "/api/v1/sessions"
      expect(response).to have_http_status(:not_found)
    end
  end
end

File: test/auth.spec.ts
import {INestApplication} from "@nestjs/common";
import jwt from "jsonwebtoken";
import {afterAll, beforeAll, beforeEach, describe, expect, it} from "vitest";

import {DataSource} from "typeorm";
import {User} from "../src/database/entities/user.entity";
import {createApp, http, HttpClient, resetDatabase} from "./helpers/app";
import {authHeader, registerAndLogin} from "./helpers/fixtures";

describe("auth", () => {
    let app: INestApplication;
    let client: HttpClient;
    let dataSource: DataSource;

    beforeAll(async () => {
        app = await createApp();
        client = http(app);
        dataSource = app.get(DataSource);
    });

    beforeEach(async () => {
        await resetDatabase(dataSource);
    });

    afterAll(async () => {
        await app.close();
    });

    it("register returns jwt and uuid without email", async () => {
        const response = await client.post("/api/v1/auth/register").send({username: "Alice", password: "secret123"});
        expect(response.status).toBe(201);
        const userId = response.body.data.user.id as string;
        expect(userId).toHaveLength(36);
        expect(userId).toMatch(/^[0-9a-f-]+$/);
        expect(response.body.data.user.username).toBe("alice");
        expect(response.body.data.token).toBeTruthy();
        expect(response.body.data.user.email).toBeUndefined();
    });

    it("register ignores email field", async () => {
        const response = await client.post("/api/v1/auth/register").send({username: "bob", password: "secret123", email: "bob@example.com"});
        expect(response.status).toBe(201);
        expect(response.body.data.user.email).toBeUndefined();
    });

    it("register rejects duplicate username case-insensitively", async () => {
        await registerAndLogin(client);
        const response = await client.post("/api/v1/auth/register").send({username: "ALICE", password: "secret123"});
        expect(response.status).toBe(422);
        expect(response.body.error.code).toBe("validation_error");
        expect(response.body.error.message).toBe("使用者名稱已被使用");
    });

    it("register rejects short password", async () => {
        const response = await client.post("/api/v1/auth/register").send({username: "alice", password: "short"});
        expect(response.status).toBe(422);
        expect(response.body.error.code).toBe("validation_error");
        expect(response.body.error.message).toBe("密碼至少需要 8 個字元");
    });

    it("login returns token for correct password and case-insensitive username", async () => {
        await registerAndLogin(client);
        const response = await client.post("/api/v1/auth/login").send({username: "ALICE", password: "secret123"});
        expect(response.status).toBe(200);
        expect(response.body.data.user.username).toBe("alice");
        expect(response.body.data.user.id).toHaveLength(36);
        expect(response.body.data.token).toBeTruthy();
    });

    it("login rejects wrong password and unknown username", async () => {
        await registerAndLogin(client);

        const wrong = await client.post("/api/v1/auth/login").send({username: "alice", password: "wrong-password"});
        expect(wrong.status).toBe(401);
        expect(wrong.body.error.code).toBe("invalid_credentials");
        expect(wrong.body.error.message).toBe("使用者名稱或密碼不正確");
        expect(wrong.body.error.request_id).toBeTruthy();

        const unknown = await client.post("/api/v1/auth/login").send({username: "nobody", password: "secret123"});
        expect(unknown.status).toBe(401);
        expect(unknown.body.error.code).toBe("invalid_credentials");
    });

    it("session endpoints are absent", async () => {
        expect((await client.delete("/api/v1/auth/logout")).status).toBe(404);
        expect((await client.get("/api/v1/sessions")).status).toBe(404);
    });

    it("me returns current user", async () => {
        const fixture = await registerAndLogin(client);
        const response = await client.get("/api/v1/me").set(authHeader(fixture.token));
        expect(response.status).toBe(200);
        expect(response.body.data.id).toBe(fixture.userId);
        expect(response.body.data.username).toBe("alice");
        expect(response.body.data.timezone).toBe("Asia/Hong_Kong");
        expect(response.body.data.currency).toBe("HKD");
    });

    it("me rejects missing, malformed and forged tokens", async () => {
        expect((await client.get("/api/v1/me")).status).toBe(401);

        const malformed = await client.get("/api/v1/me").set({Authorization: "Bearer not-a-jwt"});
        expect(malformed.status).toBe(401);
        expect(malformed.body.error.code).toBe("unauthorized");

        const fixture = await registerAndLogin(client);
        const forged = jwt.sign({user_id: fixture.userId, iat: 1_700_000_000}, "forged-secret", {
            algorithm: "HS256",
        });
        const response = await client.get("/api/v1/me").set({Authorization: `Bearer ${forged}`});
        expect(response.status).toBe(401);
    });

    it("me rejects token when user no longer exists", async () => {
        const fixture = await registerAndLogin(client);
        await dataSource.getRepository(User).delete({id: fixture.userId});
        const response = await client.get("/api/v1/me").set(authHeader(fixture.token));
        expect(response.status).toBe(401);
    });

    it("the same token works for multiple requests", async () => {
        const fixture = await registerAndLogin(client);
        const first = await client.get("/api/v1/me").set(authHeader(fixture.token));
        expect(first.status).toBe(200);
        const id = first.body.data.id;
        const second = await client.get("/api/v1/me").set(authHeader(fixture.token));
        expect(second.body.data.id).toBe(id);
    });

    it("update password success and login afterwards", async () => {
        const fixture = await registerAndLogin(client);
        const response = await client.patch("/api/v1/me/password").set(authHeader(fixture.token)).send({
            password_challenge: "secret123",
            password: "newsecret123",
            password_confirmation: "newsecret123",
        });
        expect(response.status).toBe(200);
        expect(response.body.data.username).toBe("alice");

        const login = await client.post("/api/v1/auth/login").send({username: "alice", password: "newsecret123"});
        expect(login.status).toBe(200);
        expect(login.body.data.token).toBeTruthy();

        const old = await client.post("/api/v1/auth/login").send({username: "alice", password: "secret123"});
        expect(old.status).toBe(401);
    });

    it("update password rejects bad input", async () => {
        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);

        const wrong = await client.patch("/api/v1/me/password").set(headers).send({password_challenge: "wrong-password", password: "newsecret123"});
        expect(wrong.status).toBe(422);
        expect(wrong.body.error.code).toBe("invalid_current_password");
        expect(wrong.body.error.message).toBe("目前密碼不正確");

        const missingChallenge = await client.patch("/api/v1/me/password").set(headers).send({password: "newsecret123"});
        expect(missingChallenge.status).toBe(422);
        expect(missingChallenge.body.error.code).toBe("invalid_current_password");

        const short = await client.patch("/api/v1/me/password").set(headers).send({password_challenge: "secret123", password: "short"});
        expect(short.status).toBe(422);
        expect(short.body.error.code).toBe("validation_error");
        expect(short.body.error.message).toBe("密碼至少需要 8 個字元");

        const mismatch = await client.patch("/api/v1/me/password").set(headers).send({
            password_challenge: "secret123",
            password: "newsecret123",
            password_confirmation: "different123",
        });
        expect(mismatch.status).toBe(422);
        expect(mismatch.body.error.code).toBe("validation_error");

        const missingNew = await client.patch("/api/v1/me/password").set(headers).send({password_challenge: "secret123"});
        expect(missingNew.status).toBe(422);
        expect(missingNew.body.error.code).toBe("validation_error");

        const login = await client.post("/api/v1/auth/login").send({username: "alice", password: "secret123"});
        expect(login.status).toBe(200);
    });

    it("update password requires a token", async () => {
        await registerAndLogin(client);
        const response = await client.patch("/api/v1/me/password").send({password_challenge: "secret123", password: "newsecret123"});
        expect(response.status).toBe(401);
        expect(response.body.error.code).toBe("unauthorized");
    });
});
